### DISCUSSION TITLE
Build script causes permission errors during build

### DISCUSSION THREAD

**Discussion by DrRuhe** [ipetkov/crane#189] [Q&A] [answered]
Hi, I am trying to package [samply](https://github.com/mstange/samply) using crane. My current Progress is located [in a fork](https://github.com/DrRuhe/samply/tree/main) of that repo. 

Sadly I am getting build errors pointing to the custom build commands of libc and memchr: 
```
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking sources
unpacking source archive /nix/store/8n7vs04zypwrl4chxgmf7dapkch98yqz-source
source root is source
@nix { "action": "setPhase", "phase": "patchPhase" }
patching sources
Executing configureCargoCommonVars
copying cargo artifacts from /nix/store/yjv7iqq139arf84grpl4n9wf0yy4n6kl-cargo-package-deps-0.0.1/target to target
@nix { "action": "setPhase", "phase": "configurePhase" }
configuring
will append /build/source/.cargo-home/config.toml with contents of /nix/store/1j8f3b068qkgpbaa8zcn51aarp0hqyx2-vendor-cargo-deps/config.toml
default configurePhase, nothing to do
@nix { "action": "setPhase", "phase": "buildPhase" }
building
++ command cargo --version
cargo 1.65.0 (4bc8f24d3 2022-10-20)
++ command cargo build --profile release --message-format json-render-diagnostics --target x86_64-unknown-linux-gnu
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /build/source/samply-mac-preload/Cargo.toml
workspace: /build/source/Cargo.toml
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /build/source/samply/Cargo.toml
workspace: /build/source/Cargo.toml
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /build/source/tools/query_api/Cargo.toml
workspace: /build/source/Cargo.toml
   Compiling cfg-if v1.0.0
   Compiling libc v0.2.138
   Compiling memchr v2.5.0
   Compiling bitflags v1.3.2
   Compiling pin-project-lite v0.2.9
   Compiling once_cell v1.16.0
error: failed to run custom build command for `libc v0.2.138`

Caused by:
  could not execute process `/build/source/target/release/build/libc-0060f76ec4165424/build-script-build` (never executed)

Caused by:
  Permission denied (os error 13)
warning: build failed, waiting for other jobs to finish...
error: failed to run custom build command for `memchr v2.5.0`

Caused by:
  could not execute process `/build/source/target/release/build/memchr-7d105e9cb14f44f9/build-script-build` (never executed)

Caused by:
  Permission denied (os error 13)
```

I looked for examples/documentation for this but could not find anything relevant. I'd appreciate any feedback regarding my attempts to package it and hints as to how this issue might be solved.

**Comment by ipetkov** [comment, marked as answer]
Hi @DrRuhe! 

What platform are you trying to build the project on? I see that the flake sets `cargoExtraArgs = "--target x86_64-unknown-linux-gnu"` yet it also aims to support darwin and linux builds. Judging from the attempt at building `libc` in the logs, it _feels like_ a failed cross compile (i.e. cargo produces an executable for a platform different than the host trying to execute it).

If you are looking to set up an actual cross-compiled target [this example may be useful](https://github.com/ipetkov/crane/blob/59b31b41a589c0a65e4a1f86b0e5eac68081468b/examples/cross-rust-overlay/flake.nix)

**Reply by DrRuhe** [reply]
Hi @ipetkov, 

Thanks for the quick reply! I thought this error was unrelated to the target flag thing. I am building on x86_64-unknown-linux-gnu, so I wonder what could cause cross-compilation issues.

I tried building it using `nix run .#` in the project root, so I think it should only evaluate and build for my system. 

I should have mentioned my intentions behind the target flag in the original post. So when trying to build it without the added target flag and a `flake-utils.lib.eachDefaultSystem` I got an error stating `error: language item required, but not found: 'eh_personality'`. From researching this online, I concluded that it could be fixed by adding the target flag (I don't know the exact reasoning but hey, the error disappeared).

Therefore I added the --target flag to the cargoExtraArgs.
So from there, I set up the  `flake-utils.lib.eachSystem [ "aarch64-darwin" "x86_64-darwin" "aarch64-linux" "x86_64-linux" ]` since samply only supports these systems (reading this, I am no longer sure about the aarch64 platform support).

My intention was to build a switch-like function that would take the system as arg and return the appropriate target triple. But first I wanted to get it to work using just my target, which brought me to the aforementioned error.

**Reply by ipetkov** [reply]
> I got an error stating `error: language item required, but not found: 'eh_personality'`. 

Generally I know this error can come up when building in a `#![no_std]` environment. Do you happen to have such a target in the workspace?

**Reply by DrRuhe** [reply]
Yes, turns out the `samply-mac-preload` library has a `#![no_std]` annotation. But shouldn't `#![no_std]` build just fine when std is around? I am having trouble understanding why that would cause the `eh_personality` stuff...

**Reply by DrRuhe** [reply]
Running without the added target flag gives this: 
```
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking sources
unpacking source archive /nix/store/8n7vs04zypwrl4chxgmf7dapkch98yqz-source
source root is source
@nix { "action": "setPhase", "phase": "patchPhase" }
patching sources
Executing configureCargoCommonVars
copying cargo artifacts from /nix/store/yjv7iqq139arf84grpl4n9wf0yy4n6kl-cargo-package-deps-0.0.1/target to target
@nix { "action": "setPhase", "phase": "configurePhase" }
configuring
will append /build/source/.cargo-home/config.toml with contents of /nix/store/1j8f3b068qkgpbaa8zcn51aarp0hqyx2-vendor-cargo-deps/config.toml
default configurePhase, nothing to do
@nix { "action": "setPhase", "phase": "buildPhase" }
building
++ command cargo --version
cargo 1.65.0 (4bc8f24d3 2022-10-20)
++ command cargo build --profile release --message-format json-render-diagnostics
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /build/source/samply-mac-preload/Cargo.toml
workspace: /build/source/Cargo.toml
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /build/source/samply/Cargo.toml
workspace: /build/source/Cargo.toml
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /build/source/tools/query_api/Cargo.toml
workspace: /build/source/Cargo.toml
   Compiling samply-symbols v0.18.0 (/build/source/samply-symbols)
   Compiling fxprof-processed-profile v0.4.0 (/build/source/fxprof-processed-profile)
   Compiling perfrecord-mach-ipc-rendezvous v0.2.1 (/build/source/perfrecord-mach-ipc-rendezvous)
   Compiling gecko_profile v0.4.0 (/build/source/gecko_profile)
   Compiling samply-mac-preload v0.1.0 (/build/source/samply-mac-preload)
error: language item required, but not found: `eh_personality`
  |
  = note: this can occur when a binary crate with `#![no_std]` is compiled for a target where `eh_personality` is defined in the standard library
  = help: you may be able to compile for a target that doesn't need `eh_personality`, specify a target with `--target` or in `.cargo/config`

error: could not compile `samply-mac-preload` due to previous error
warning: build failed, waiting for other jobs to finish...
```

Which is in line with what you said. So If I understand correctly, the eh_personality thing must be defined in the lib/bin but since this is no_std it is not there and then rustc does not know what to do?

**Reply by ipetkov** [reply]
It looks like `samply-mac-preload` is meant to compile on darwin, I wonder if the `eh_personality` stuff is expected to come from the platform-dependent SDKs, but when built under Linux it fails with that error. My suggestion is to avoid building that crate (e.g. via `cargo build --workspace --exclude samply-mac-preload`) on non-darwin platforms

**Reply by DrRuhe** [reply]
I got the package to build 🥳
So, instead of excluding only `sampy-mac-preload` I thought "shouldn't cargo just ignore it anyway when I try building only the samply binary on a non-darwin platform?". So now I just use `--bin samply` as a cargo extra arg and it only builds the required libs, thus excluding `samply-mac-preload`.

Now I am just having trouble getting to actually run this:
```
$ nix run .#
error: unable to execute '/nix/store/2pdnvpfx7sgy691xkfv60pgp9hzjq7xn-cargo-package-0.0.1/bin/cargo-package': No such file or directory
```

So how can I change the following, such that It works the same "as usual"?
```
samply = craneLib.buildPackage {
          inherit cargoArtifacts src buildInputs;
          cargoExtraArgs = "--bin samply";
        };
```

This boils down to the difference in handling crates as opposed to workspaces. I am having trouble finding documentation regarding this.

**Reply by ipetkov** [reply]
@DrRuhe `nix run` assumes that if you have a derivation named `foo` then it can run an executable at `${foo}/bin/foo`. In this case it looks like it is trying to run a binary from `./bin/cargo-package` which doesn't exist. If you tweak the executable path (and I believe `flake-utils` has a `mkApp` utility to make that easier) it should fix the `nix run .#` invocation

**Reply by DrRuhe** [reply]
Ok, so after some more tinkering, I got it to work as I want it to. I solved it by adding a `pname` to the `craneLib.buildPackage` invocation like this:

```
samply = craneLib.buildPackage {
          inherit cargoArtifacts src buildInputs;
          pname = "samply";
          cargoExtraArgs = "--bin samply";
};
```

I took the inspiration to try this from the [fedimint flake.nix](https://github.com/fedimint/fedimint/blob/master/flake.nix).

**Reply by DrRuhe** [reply]
Thanks for the help :)
